client/src/document/hooks.ts

```diff
@@ -8,3 +11,79 @@ export function useDocumentURL() {
   // pathname with the document's mdn_url.
   return url.endsWith("/") ? url.substring(0, url.length - 1) : url;
 }
+
+export function useCopyExamplesToClipboard(doc: Doc | undefined) {
+  React.useEffect(() => {
+    if (!doc) {
+      return;
+    }
+    // '#content pre' works but also selects <pre>
+    // tags without class="brush:*", like class="notranslate"
+    // which represent 'formal_syntax' <pre> content, and
+    // syntax highlight isn't targeting <pre class="notranslate">
+    // its getting $(pre[class*=brush]), this is in regards to wrapping
+    // <pre> tags with a <div> to absolutely position the
+    // copy-to-clipboard <button> relatively in the newly added parent div,
+    // instead of selecting #content pre which grabs
+    // class="notranslate" formal syntax tags, just grab
+    // the <pre class="brush"> tags which are injected into
+    // a `<div class="code-example">` from syntax highlighter
+    [...document.querySelectorAll("div.code-example pre")].forEach(
+      (element) => {
+        const userMessage = document.createElement("span");
+        const button = document.createElement("button");
+        const span = document.createElement("span");
+        userMessage.textContent = "Copied!";
+        span.textContent = "Copy to Clipboard";
+
+        const wrapper: HTMLElement | null = element.parentElement;
+
+        userMessage.setAttribute("class", "user-message");
+        userMessage.setAttribute("aria-hidden", "true");
+        button.setAttribute("aria-hidden", "false");
+        button.setAttribute("type", "button");
+        button.setAttribute("class", "copy-icon");
+        span.setAttribute("class", "visually-hidden");
+
+        button.appendChild(span);
+        wrapper?.appendChild(button);
+        wrapper?.appendChild(userMessage);
+
+        button.onclick = () => {
+          userMessage.classList.add("show");
+          userMessage.setAttribute("aria-hidden", "false");
+          copyToClipboard(element);
+          button.classList.add("copied");
+          userMessage.style.top = "52px";
+
+          setTimeout(() => {
+            userMessage.classList.remove("show");
+            userMessage.setAttribute("aria-hidden", "true");
+          }, 1000);
```

Comment:
Very nice recommendation. I totally forgot to even think about if `navigator.clipboard.writeText` would throw an error if the API indeed fails to copy. I think adding `try/catch` will be another layer of defense and will help us handle when it *doesn't work* since we are aware that failures can happen. 

> I can't clearly see how it could fail but it can, it always can.

Right! I can't clearly think how it could fail in the copying process as I'm not sure how the `navigator.clipboard.writeText` is implemented. But as you said, one thing at a time :)

Reply:
@peterbe I think making the `copyToClipboard` function return a boolean and assigning the return value to a variable to use in a check within the custom hook is the way to go. Since `navigator.clipboard.writeText` is "permission-aware" ie uses the [Permissions API](https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API#concepts_and_usage) and returns `Promise<void>`, we can better understand when the copy-to-clipboard utility wouldn't allow a copy and fails.

> The "clipboard-write" permission of the Permissions API, is granted automatically to pages when they are in the active tab. [Clipboard.writeText()](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/writeText)

Since we know all of this about `Clipboard.writeText` we can provide some better failure defense within the `copyToClipboard` function since `navigator.clipboard.writeText` is indeed returning a Promise that will resolve when the Clipboard contents have been set/updated and will reject if the caller doesn't have permissions (from Permissions object that Permissions API checks) to write to the clipboard.